File: pay_contributor.py
import xpring
import requests
import re
import os

def pay_xrp_testnet(wallet_seed, address, amount):
    # Create a wallet instance using the seed
    wallet = xpring.Wallet.from_seed(wallet_seed)

    # The XRP testnet
    url = 'test.xrp.xpring.io:50051'
    client = xpring.Client.from_url(url)

    # Create a transaction
    txn = client.send(wallet, address, str(amount))

    # Submit txn to the network
    res = client.submit(txn)

    return res

def get_address_from_payid(payid, network, environment):
    # Convert the PayID into a URL e.g.
    # pay$username.github.io -> https://username.github.io/pay
    local_part, domain = payid.split('$')
    url = f'https://{domain}/{local_part}'
    response = requests.get(url)
    response.raise_for_status()

    data = response.json()

    # Look for an address that matches the network
    # and environment we want to use
    for address in data['addresses']:
        if address['paymentNetwork'] == network and \
           address['environment'] == environment:
            return address['addressDetails']['address']

def find_all_payids(msg):
    return re.findall(r'(\S+\$\S+\.\S+)', msg)

if __name__ == '__main__':
    print("running pay contributor")

    # Get out intputs from the environment
    commitmsg = os.environ['INPUT_COMMIT_LOG']
    secret = os.environ['INPUT_WALLET_SECRET']
    amount = int(os.environ['INPUT_AMOUNT'])
    max_payout = int(os.environ['INPUT_MAX_PAYOUT'])

    print("commit message:", commitmsg)
    payids = find_all_payids(commitmsg)
    print("Payids found:", payids)

    # Ensure the total max payout is respected
    amount = int(min(amount, max_payout / len(payids)))

    # Pay each payid
    for payid in payids:
        address = get_address_from_payid(payid, 'XRPL', 'TESTNET')
        if address:
            print(f"Address found for {payid} is {address}, paying {amount}")
            pay_xrp_testnet(secret, address, amount)
        else:
            print("No address found for payid:", payid)

File: pay_contributor.js
const { PayIdClient, Wallet,
	XrpClient, XrplNetwork,
	XrpPayIdClient, XpringClient } = require('xpring-js')

const core = require('@actions/core')
const github = require('@actions/github');

require('dotenv').config()

async function run() {

    // Get our parameters from the environment
    const env = process.env
    const wallet_seed = process.env.INPUT_WALLET_SECRET
    const environment = process.env.INPUT_ENVIRONMENT.toLowerCase()
    const server = process.env.INPUT_SERVER
    const commitmsg = process.env.INPUT_COMMIT_LOG
    const amount = process.env.INPUT_AMOUNT
    const max_payout = process.env.INPUT_MAX_PAYOUT

    // If not commit message, then try to work out open opener of pull request
    const token = core.getInput('repo_token')
    const octokit = github.getOctokit(token)

    const context = github.context
    console.log(JSON.stringify(context, null, 4))

    
    // Abort if we don't have a wallet secret
    if (wallet_seed === undefined) {
	console.log("Missing INPUT_WALLET_SECRET")
	process.exit(1)
    }

    // Instantiate instance of a wallet with seed
    const wallet = Wallet.generateWalletFromSeed(
	wallet_seed,
    )

    // Create the clients we need
    const xrpClient = new XrpClient(server, environment)
    const payIdClient = new XrpPayIdClient(environment)
    const xpringClient = new XpringClient(payIdClient, xrpClient)

    // Find all payids in the commit message
    const payIds = commitmsg.match(/(\S+\$\S+\.\S+)/g)

    // If we have no payids found then exit with success
    const num = payIds.length
    if (num < 1) {
	console.log("No PayIDs found")
	process.exit(0)
    }

    // Calculate the amount to pay, paying each evenly
    const payid_amount = Math.floor(Math.min(amount, max_payout / num))

    // Make each payment and await success
    for(let i=0; i<num; i++) {
	let payId = payIds[i]
	console.log(`Paying ${payId} amount ${payid_amount}`)
	try {
	    const transactionHash = await xpringClient.send(payid_amount,
							    payId,
							    wallet)
	    console.log(transactionHash)
	} catch(e) {
	    console.log("Could not pay", payId, e)
	}
	
    }
}

run();
